src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -293,21 +295,28 @@ private bool StartCore(ProcessStartInfo startInfo)
                 throw new Win32Exception(Interop.Error.ENOENT.Info().RawErrno);
             }
 
-            // Invoke the shim fork/execve routine.  It will create pipes for all requested
-            // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
-            // descriptors, and execve to execute the requested process.  The shim implementation
-            // is used to fork/execve as executing managed code in a forked process is not safe (only
-            // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
-            Interop.Sys.ForkAndExecProcess(
-                    filename, argv, envp, cwd,
-                    startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
-                    out childPid,
-                    out stdinFd, out stdoutFd, out stderrFd);
-
-            // Store the child's information into this Process object.
-            Debug.Assert(childPid >= 0);
-            SetProcessId(childPid);
-            SetProcessHandle(new SafeProcessHandle(childPid));
+            // Lock to avoid races with OnSigChild
+            lock (s_sigchildGate)
```

Comment:
This is really unfortunate, introducing a fairly chunky process-wide lock on every Process.Start invocation.  I think it's important to find an alternative.

Reply:
We could use a ReaderWriterLock. The _readers_ are Process.Start, the _writer_ is OnSigChild. wdyt?